cache.go

```diff
@@ -151,7 +162,14 @@ func (d DirCache) Put(ctx context.Context, name string, cert *tls.Certificate) e
 			// Don't overwrite the file if the context was canceled.
 		default:
 			newName := filepath.Join(string(d), name)
-			err = os.Rename(tmp, newName)
+			err = os.Rename(tmpKey, newName+".key")
+			if err != nil {
+				return
+			}
+			err = os.Rename(tmpCert, newName+".cert")
+			if err != nil {
+				return
```

Comment:
If this fails, should we rename the other one back? :thinking:

Reply:
While writing, I actually was thinking that it should be atomic: if either of the Renames fail, it should delete both tmp files and return an error. One of the things I noticed that initially prompted me to do this is that the cache dir was filling up with tmp files that weren't being renamed because of errors in the encoding.

Reply:
Yeah, if we can accomplish that it would be good, or at least make a best effort.